### PR TITLE
Kops controller support for OpenStack

### PR DESCRIPTION
**Pull request by zetaab**
Implements OpenStack part to #7496

### CODE REVIEW

File: pkg/nodeidentity/openstack/identify.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		if err != nil {
+			return nil, err
+		}
+		err = json.Unmarshal(bodyBytes, &meta)
+		if err != nil {
+			return nil, err
+		}
+		return &meta, nil
+	}
+	return nil, err
```
Comment (rifelpet) [inline pkg/nodeidentity/openstack/identify.go]: This `err` variable will always be `nil` here, maybe return an `errors.New()` or `fmt.Errorf()` instead?

File: pkg/nodeidentity/openstack/identify.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+// New creates and returns a nodeidentity.Identifier for Nodes running on OpenStack
+func New() (nodeidentity.Identifier, error) {
+	return &nodeIdentifier{}, nil
+}
+
+// IdentifyNode queries OpenStack for the node identity information
+func (i *nodeIdentifier) IdentifyNode(ctx context.Context, node *corev1.Node) (*nodeidentity.Info, error) {
+	metadata, err := getLocalMetadata()
```
Comment (justinsb) [inline pkg/nodeidentity/openstack/identify.go]: Ah - sorry - there's a gotcha here.  We need to get the instancegroup of `node`, which usually won't be our node.  The idea is that we can't trust the nodes to self-report (or we don't want to).

Looking at some of the existing openstack code, I think you need to do a ListInstances request and then get the `KopsInstanceGroup` metadata (nice forward thinking of adding that metadata already).  You can identify the instance using the node.Spec.ProviderID which should (hopefully) be set on the node by openstack cloudprovider.

I'm not sure if the instance metadata is trivially writeable from the instance, for example by posting to http://169.254.169.254.  If it is we might need to beef up the security somehow, but I think the ProviderID is currently unprotected anyway...

**Comment by zetaab** [comment]
Something wrong this will mark same instancegroup to all instances:

```
root@master-zone-1-3-1-calc-k8s-local:/home/debian# kubectl get nodes --show-labels
NAME                               STATUS     ROLES    AGE   VERSION   LABELS
master-zone-1-1-1-calc-k8s-local   Ready      master   48s   v1.16.0   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=f1cd44ae-eb5e-4529-9a47-2dd5f9cfdbe3,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=regionOne,failure-domain.beta.kubernetes.io/zone=zone-1,kops.k8s.io/instancegroup=master-zone-1-3,kubernetes.io/arch=amd64,kubernetes.io/hostname=master-zone-1-1-1-calc-k8s-local,kubernetes.io/os=linux,kubernetes.io/role=master,node-role.kubernetes.io/master=
master-zone-1-3-1-calc-k8s-local   Ready      master   52s   v1.16.0   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=f1cd44ae-eb5e-4529-9a47-2dd5f9cfdbe3,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=regionOne,failure-domain.beta.kubernetes.io/zone=zone-1,kops.k8s.io/instancegroup=master-zone-1-3,kubernetes.io/arch=amd64,kubernetes.io/hostname=master-zone-1-3-1-calc-k8s-local,kubernetes.io/os=linux,kubernetes.io/role=master,node-role.kubernetes.io/master=
nodes-1-calc-k8s-local             NotReady   master   12s   v1.16.0   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=f1cd44ae-eb5e-4529-9a47-2dd5f9cfdbe3,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=regionOne,failure-domain.beta.kubernetes.io/zone=zone-1,kops.k8s.io/instancegroup=master-zone-1-3,kubernetes.io/arch=amd64,kubernetes.io/hostname=nodes-1-calc-k8s-local,kubernetes.io/os=linux,kubernetes.io/role=master,node-role.kubernetes.io/master=
nodes-2-calc-k8s-local             NotReady   master   16s   v1.16.0   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=f1cd44ae-eb5e-4529-9a47-2dd5f9cfdbe3,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=regionOne,failure-domain.beta.kubernetes.io/zone=zone-1,kops.k8s.io/instancegroup=master-zone-1-3,kubernetes.io/arch=amd64,kubernetes.io/hostname=nodes-2-calc-k8s-local,kubernetes.io/os=linux,kubernetes.io/role=master,node-role.kubernetes.io/master=
```

metadata do have correct values, but kops-controller somehow is not using it correctly?

```
root@master-zone-1-3-1-calc-k8s-local:/home/debian# curl -s http://169.254.169.254/openstack/latest/meta_data.json|jq .meta.KopsInstanceGroup
"master-zone-1-3"
root@master-zone-1-2-1-calc-k8s-local:/home/debian# curl -s http://169.254.169.254/openstack/latest/meta_data.json|jq .meta.KopsInstanceGroup
"master-zone-1-2"
root@master-zone-1-1-1-calc-k8s-local:/home/debian# curl -s http://169.254.169.254/openstack/latest/meta_data.json|jq .meta.KopsInstanceGroup
"master-zone-1-1"
root@nodes-2-calc-k8s-local:/home/debian# curl -s http://169.254.169.254/openstack/latest/meta_data.json|jq .meta.KopsInstanceGroup
"nodes"
```

@justinsb any idea what is going on? is there some cache somewhere?

File: pkg/nodeidentity/openstack/identify.go (hunk 2 above)
Comment (zetaab) [reply pkg/nodeidentity/openstack/identify.go]: okay then its bigger change :) I thought this is going to be easy!

**Comment by zetaab** [comment]
```
root@master-zone-1-3-1-calc-k8s-local:/home/debian# kubectl get nodes --show-labels
NAME                               STATUS   ROLES    AGE    VERSION   LABELS
master-zone-1-1-1-calc-k8s-local   Ready    master   109s   v1.16.0   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=f1cd44ae-eb5e-4529-9a47-2dd5f9cfdbe3,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=regionOne,failure-domain.beta.kubernetes.io/zone=zone-1,kops.k8s.io/instancegroup=master-zone-1-1,kubernetes.io/arch=amd64,kubernetes.io/hostname=master-zone-1-1-1-calc-k8s-local,kubernetes.io/os=linux,kubernetes.io/role=master,node-role.kubernetes.io/master=
master-zone-1-2-1-calc-k8s-local   Ready    master   105s   v1.16.0   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=f1cd44ae-eb5e-4529-9a47-2dd5f9cfdbe3,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=regionOne,failure-domain.beta.kubernetes.io/zone=zone-1,kops.k8s.io/instancegroup=master-zone-1-2,kubernetes.io/arch=amd64,kubernetes.io/hostname=master-zone-1-2-1-calc-k8s-local,kubernetes.io/os=linux,kubernetes.io/role=master,node-role.kubernetes.io/master=
master-zone-1-3-1-calc-k8s-local   Ready    master   61s    v1.16.0   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=f1cd44ae-eb5e-4529-9a47-2dd5f9cfdbe3,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=regionOne,failure-domain.beta.kubernetes.io/zone=zone-1,kops.k8s.io/instancegroup=master-zone-1-3,kubernetes.io/arch=amd64,kubernetes.io/hostname=master-zone-1-3-1-calc-k8s-local,kubernetes.io/os=linux,kubernetes.io/role=master,node-role.kubernetes.io/master=
nodes-1-calc-k8s-local             Ready    node     26s    v1.16.0   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=f1cd44ae-eb5e-4529-9a47-2dd5f9cfdbe3,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=regionOne,failure-domain.beta.kubernetes.io/zone=zone-1,kops.k8s.io/instancegroup=nodes,kubernetes.io/arch=amd64,kubernetes.io/hostname=nodes-1-calc-k8s-local,kubernetes.io/os=linux,kubernetes.io/role=node,node-role.kubernetes.io/node=
nodes-2-calc-k8s-local             Ready    node     29s    v1.16.0   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=f1cd44ae-eb5e-4529-9a47-2dd5f9cfdbe3,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=regionOne,failure-domain.beta.kubernetes.io/zone=zone-1,kops.k8s.io/instancegroup=nodes,kubernetes.io/arch=amd64,kubernetes.io/hostname=nodes-2-calc-k8s-local,kubernetes.io/os=linux,kubernetes.io/role=node,node-role.kubernetes.io/node=
```

**Comment by zetaab** [comment]
@justinsb could you review this, we could get one step closer to 1.16 alpha

File: pkg/nodeidentity/openstack/identify.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+// New creates and returns a nodeidentity.Identifier for Nodes running on OpenStack
+func New() (nodeidentity.Identifier, error) {
+	env, err := openstack.AuthOptionsFromEnv()
+	if err != nil {
+		return nil, err
+	}
+
+	region := os.Getenv("OS_REGION_NAME")
+	if region == "" {
+		return nil, fmt.Errorf("Unable to find region")
```
Comment (justinsb) [inline pkg/nodeidentity/openstack/identify.go]: Nit: A message like "OS_REGION_NAME not set" might be more self-documenting

File: pkg/nodeidentity/openstack/identify.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
+		return nil, fmt.Errorf("providerID was not set for node %s", node.Name)
+	}
+	if !strings.HasPrefix(providerID, "openstack://") {
+		return nil, fmt.Errorf("providerID %q not recognized for node %s", providerID, node.Name)
+	}
+
+	instanceID := strings.TrimPrefix(providerID, "openstack://")
+	// instanceid looks like its openstack:/// but no idea is that really correct like that?
```
Comment (justinsb) [inline pkg/nodeidentity/openstack/identify.go]: This is probably my fault ... back in the day (on AWS) I thought we'd put the region or zone in here for WAN clusters, so there's an extra slash in here.  Sorry!

**Pull request merged**